### PR TITLE
Fix memory leak in circular maps

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

Resolves: #1795

## Description

The function _ebpf_core_circular_map_peek_or_pop is reachable both via ebpf_map_pop_entry and ebpf_map_find_entry, but the memory is only freed in the ebpf_map_pop_entry path. The fix is to move it into _ebpf_core_circular_map_peek_or_pop so that the memory is freed on both paths.

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
Need to figure out how to handle the annotation correctly. Memory free by ebpf_epoch_free is still valid until the call returns, but I am not sure how to annotate that.

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -188,8 +188,15 @@ _ebpf_core_circular_map_peek_or_pop(_Inout_ ebpf_core_circular_map_t* map, bool
             map->end = new_end;
         }
     }
+    if (pop) {
+        // The return_value is not freed until the current epoch is retired.
+        ebpf_epoch_free(return_value);
+    }
 Done:
+#pragma warning(push)
+#pragma warning(disable : 6001) // Using uninitialized memory 'return_value'.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:197]: I'd suggest adding `_Ret_maybenull_` to line 162 but that can be in a separate PR

**Review by dthaler — APPROVED** [review]

**Review by dv-msft — APPROVED** [review]
